### PR TITLE
Fix missing musttail for dynamic replacement calls.

### PR DESCRIPTION
**Pull request by varungandhi-apple**
When we make a tail call from swifttailcc → swifttailcc, the call should be marked musttail even when dynamic replacement is involved. However, this wasn't handled earlier, leading to an assertion failure when I tried to turn on musttail verification. (See https://github.com/apple/swift/pull/36805#issuecomment-844418793)

Fixes rdar://78284346 .

### CODE REVIEW

**Comment by varungandhi-apple** [comment]
@swift-ci test

File: lib/IRGen/GenDecl.cpp [hunk 1]
```diff
@@ -2955,6 +2955,10 @@ void IRGenModule::emitDynamicReplacementOriginalFunctionThunk(SILFunction *f) {
       FunctionPointer(fnType, typeFnPtr, authInfo, signature)
           .getAsFunction(IGF),
       forwardedArgs);
+  if (Res->getCallingConv() == llvm::CallingConv::SwiftTail &&
```
Comment (aschwaighofer) [inline lib/IRGen/GenDecl.cpp:2958]: I _think_ you can just write `if (f->isAsync())`

**Comment by varungandhi-apple** [comment]
@swift-ci test

**Review by aschwaighofer — APPROVED** [review]
lgtm. Thank you!

**Comment by varungandhi-apple** [comment]
@swift-ci test macOS

**Comment by varungandhi-apple** [comment]
Revert in https://github.com/apple/swift/pull/37645

**Comment by varungandhi-apple** [comment]
Reverting was the wrong thing; the test should be x-failed on Windows. https://github.com/apple/swift/pull/37646

**Pull request merged**